### PR TITLE
Eislam/sy 998 ni heap corruption and driver crash

### PR DESCRIPTION
**Pull request by Lham42**
# Fix Pull Request Template

## Key Information

- **Linear Issue**: [SY-998]() and  [SY-999]()

## Description

Maximum 2-3 sentence description describing the changes.

This pull request makes a number of changes in order to improve stability on the driver, particularly with driver crashing. Some changes:

- eliminating unnecessary clearing of the tasks vector in the task manager which causes 
- ensuring scanner is destroyed after external threads which use it have finished their thread of execution
- added a preprocessor macro to handle constructing ni factory which ensures that driver does not crash even when using a mock library

## Basic Readiness Checklist

- [x] I have performed a self-review of my code.
- [x] I have added sufficient regression tests to cover the changes to CI.
- [x] I have added relevant tests to cover the changes or exposing bugs.
- [x] I have verified code coverage targets are met.

## Additional Notes
- [x] These changes deal with concurrency.
- [ ] These changes affect UI.

## Manual QA Additions

- [x] I have updated the [Release Candidate](/.github/PULL_REQUEST_TEMPLATE/rc.md) template
  with necessary manual QA steps to test my change.

## Reviewer Checklist
- [ ] Sufficient test coverage of new additions.
- [ ] Verified all steps in readiness checklists.
- [ ] UI changes have been tested.
- [ ] style and formatting is consistent.
- [ ] Reviewed any relevant changes to concurrent code for safety. 
- [ ] Sufficient comments and clarity of code.

### CODE REVIEW

File: driver/BUILD.bazel [hunk 1]
```diff
@@ -47,10 +54,14 @@ cc_binary(
     srcs = select({
         "//conditions:default":["main.cpp"],
     }),
+    # copts = select({
```
Comment (emilbon99) [inline driver/BUILD.bazel:57]: Can we remove this commented out stuff?

File: driver/ni/scanner.cpp [hunk 2]
```diff
@@ -62,14 +62,21 @@ ni::Scanner::Scanner(
...
     ni::NiSysCfgInterface::CloseHandle(this->resources_handle);
     ni::NiSysCfgInterface::CloseHandle(this->session);
+    if(this->scan_thread && scan_thread->joinable()) scan_thread->join();
 }
 
 void ni::Scanner::scan() {
+    if(!this->ok_state) return;
     NISysCfgResourceHandle resource = NULL;
 
     // first find hardware
```
Comment (emilbon99) [inline driver/ni/scanner.cpp:82]: Useless comment

File: driver/task/manager.cpp [hunk 3]
```diff
@@ -107,7 +107,6 @@ freighter::Error task::Manager::stop() {
         task->stop();
         LOG(INFO) << "[driver] task " << task->name() << " stopped";
     }
-    tasks.clear();
```
Comment (emilbon99) [inline driver/task/manager.cpp:110]: Why did you clear this?

**Review by emilbon99 — APPROVED** [review]
Should this be merged into main?

File: driver/task/manager.cpp (hunk 3 above)
Comment (Lham42) [reply driver/task/manager.cpp:110]: Are you asking why we cleared - i think before it had seemed to fix  an error we received on shutdown but did not address the root of the problem. It then became the cause of one of the heap corruption issues we are currently facing.

**Pull request merged**